flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/AsyncWaitOperator.java

```diff
@@ -209,41 +176,34 @@ else if (element.isLatencyMarker()) {
 	}
 
 	@Override
-	public void processElement(StreamRecord<IN> element) throws Exception {
-		final StreamRecordQueueEntry<OUT> streamRecordBufferEntry = new StreamRecordQueueEntry<>(element);
+	public void processElement(final StreamRecord<IN> element) throws Exception {
+		// add element first to the queue
+		final ResultFuture<OUT> entry = addToWorkQueue(element);
+
+		final ResultHandler resultHandler = new ResultHandler(element, entry);
 
+		// register a timeout for the entry if timeout is configured
 		if (timeout > 0L) {
-			// register a timeout for this AsyncStreamRecordBufferEntry
-			long timeoutTimestamp = timeout + getProcessingTimeService().getCurrentProcessingTime();
-
-			final ScheduledFuture<?> timerFuture = getProcessingTimeService().registerTimer(
-				timeoutTimestamp,
-				new ProcessingTimeCallback() {
-					@Override
-					public void onProcessingTime(long timestamp) throws Exception {
-						userFunction.timeout(element.getValue(), streamRecordBufferEntry);
-					}
-				});
-
-			// Cancel the timer once we've completed the stream record buffer entry. This will remove
-			// the register trigger task
-			streamRecordBufferEntry.onComplete(
-				(StreamElementQueueEntry<Collection<OUT>> value) -> {
-					timerFuture.cancel(true);
-				},
-				executor);
-		}
+			final long timeoutTimestamp = timeout + getProcessingTimeService().getCurrentProcessingTime();
 
-		addAsyncBufferEntry(streamRecordBufferEntry);
+			final ScheduledFuture<?> timeoutTimer = getProcessingTimeService().registerTimer(
+					timeoutTimestamp,
+					timestamp -> userFunction.timeout(element.getValue(), resultHandler));
 
-		userFunction.asyncInvoke(element.getValue(), streamRecordBufferEntry);
+			resultHandler.setTimeoutTimer(timeoutTimer);
```

Comment:
I have potential worries of setting timer delay here. I mean if above `registerTimer` happens before this setting, and during `ResultHandler#complete` if `mailboxExecutor.execute` also happens immediately, then the un-set timers would not be canceled any more.  

I know it would not happen atm because `mailboxExecutor.execute` just inserts into queue and would not execute right now. Maybe my worry could be ignored.

Reply:
I think it's still possible that a timer callback can be executed later: it may happen, that the timer has triggered and posted a letter to the mailbox, but that letter would be executed later than the letter corresponding to proper result completion. Cancelling the `timeoutTimer` won't have much effect in that case.

Reply:
Hm that's definitively not what I expected from timers and I'm pretty sure it's hard for end-users to understand. My basic idea is: if I cancel a timer from 1 thread, then it will not execute in the same thread.

I would suggest hotfixing that in general by adding a safeguard at the start of the timer letter. WDYT? @1u0

Reply:
Allowing cancellation of timer callbacks that are already in mailbox queue may help in this situation, but still won't have strict guarantees in general (as it was before with multithreaded code).

Even with a hotfix, you may still need to somehow track/allow only one result (either successfull or failed) per `ResultHandler`. With current implementation, an ill written `AsyncFunction` may call multiple times `resultHandler.complete()` (on the same ``resultHandler` instance) and every such call would post a new mailbox letter.

Reply:
Added a guard now in `ResultHandler`.